staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/kms/kms.go

```diff
@@ -0,0 +1,177 @@
+// used decrypted DEKs in memory.
+func NewKMSTransformer(kmsService Service, cacheSize int) (value.Transformer, error) {
+	if cacheSize == 0 {
+		cacheSize = defaultCacheSize
+	}
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		return nil, err
+	}
+	return &kmsTransformer{
+		kmsService:   kmsService,
+		transformers: cache,
+		cacheSize:    cacheSize,
+	}, nil
+}
+
+// TransformFromStorage decrypts data encrypted by this transformer using envelope encryption.
+func (t *kmsTransformer) TransformFromStorage(data []byte, context value.Context) ([]byte, bool, error) {
+	fmt.Println("KMS T: FROM")
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	// Length of key can be encoded in 16 bits.
+	keyLen := int(binary.BigEndian.Uint16(data[:2]))
+	if keyLen+2 > len(data) {
+		return []byte{}, false, fmt.Errorf("invalid data encountered by gkms transformer, length longer than available bytes: %q", data)
+	}
+	encKey := string(data[2 : keyLen+2])
+	encData := data[2+keyLen:]
+
+	var transformer value.Transformer
+	var kmsStale bool
+	_transformer, found := t.transformers.Get(encKey)
+	if found {
+		transformer = _transformer.(value.Transformer)
+	} else {
+		key, err := t.kmsService.Decrypt(encKey)
+		if err != nil {
+			return []byte{}, false, fmt.Errorf("error while decrypting key: %q", err)
+		}
+
+		// We need to release the read lock to prevent a deadlock
+		t.lock.RUnlock()
+		transformer, err = t.addTransformer(encKey, key)
+		t.lock.RLock()
+
+		if err != nil {
+			return []byte{}, false, err
+		}
+	}
+	kmsStale, err := t.kmsService.CheckStale(encKey)
+	if err != nil {
+		return []byte{}, false, err
+	}
+	res, transformerStale, err := transformer.TransformFromStorage(encData, context)
+	return res, (transformerStale || kmsStale), err
+}
+
+// TransformToStorage encrypts data to be written to disk using envelope encryption.
+func (t *kmsTransformer) TransformToStorage(data []byte, context value.Context) ([]byte, error) {
+	fmt.Println("KMS T: TO")
+	newKey, err := generateKey(32)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	encKey, err := t.kmsService.Encrypt(newKey)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	transformer, err := t.addTransformer(encKey, newKey)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	encKeyLen := make([]byte, 2)
+	encKeyBytes := []byte(encKey)
+	binary.BigEndian.PutUint16(encKeyLen, uint16(len(encKeyBytes)))
+
+	prefix := append(encKeyLen, encKeyBytes...)
+
+	prefixedData := make([]byte, len(prefix), len(data)+len(prefix))
+	copy(prefixedData, prefix)
+	result, err := transformer.TransformToStorage(data, context)
+	if err != nil {
+		return nil, err
+	}
+	prefixedData = append(prefixedData, result...)
+	return prefixedData, nil
+}
+
+var _ value.Transformer = &kmsTransformer{}
+
+// addTransformer inserts a new transformer to the KMS cache of DEKs for future reads.
+func (t *kmsTransformer) addTransformer(encKey string, key []byte) (value.Transformer, error) {
+	block, err := aes.NewCipher(key)
```

Comment:
Yes. THat's right. Pick a default, e.g. AES-CBC with a 256 bit key length. But, if someone wants a shorter key length of say, 192 bits they can reconfigure. The point is that its hard to predict which encryption algorithms, cipher modes and key lengths are required in a deployment. Providing this configurability lessens the need to change the code for a different encryption algorithm, mode or key length.